vpr/src/route/connection_router.cpp

```diff
@@ -372,7 +372,7 @@ void ConnectionRouter<Heap>::timing_driven_expand_cheapest(t_heap* cheapest,
         VTR_LOGV_DEBUG(router_debug_, "    Better cost to %d\n", inode);
         VTR_LOGV_DEBUG(router_debug_, "    New total cost: %g\n", new_total_cost);
         VTR_LOGV_DEBUG(router_debug_, "    New back cost: %g\n", new_back_cost);
-        VTR_LOGV_DEBUG(router_debug_, "      Setting path costs for associated node %d (from %d edge %zu)\n",
+        VTR_LOGV_DEBUG(router_debug_ && (rr_nodes_.node_type(RRNodeId(cheapest->index)) != t_rr_type::SOURCE), "      Setting path costs for associated node %d (from %d edge %zu)\n",
```

Comment:
I think it may be better to check whether the type is SOURCE instead of checking for the edge ID validity for the following reasons:

1. We assume that the only time prev_edge is invalid is when the node is of the `SOURCE `type. Thus, it may be better to raise an error if the edge ID is not valid and the node type is not SOURCE.
2. Previously, the return type of the `prev_edge` function was `RREdgeId`, but now, for performance reasons, it is a union type, and the other value in the union is initialized with `nullptr`. I am not sure if casting this to uint32 would result in an invalid `RREdgeId`(-1).

I added checks with `VTR_ASSERT_DEBUG` in the `rr graph storage` to ensure that the arrays storing edge sources and sinks are not accessed by invalid IDs.

Reply:
OK, thanks.